dev/RadioButtons/RadioButtons.cpp

```diff
@@ -7,136 +7,634 @@
     SetValue(s_ItemsProperty, items);
 
     SetDefaultStyleKey(this);
+
+    // Override normal up/down/left/right behavior -- down should always go to the next item and up to the previous.
+    // left and right should be spacial but contained to the RadioButtons control. We have to attach to PreviewKeyDown
+    // because RadioButton has a key down handler for up and down that gets called before we can intercept.
+    PreviewKeyDown({ this, &RadioButtons::OnChildKeyDown });
+    GettingFocus({ this, &RadioButtons::OnGettingFocus });
+    GotFocus({ this, &RadioButtons::OnChildGotFocus });
+    KeyDown({ this, &RadioButtons::KeyDownHandler });
+    KeyUp({ this, &RadioButtons::KeyUpHandler });
+
+    s_childHandlersProperty =
+        InitializeDependencyProperty(
+            s_childHandlersPropertyName,
+            winrt::name_of<ChildHandlers>(),
+            winrt::name_of<winrt::RadioButtons>(),
+            true /* isAttached */,
+            nullptr,
+            nullptr);
 }
 
 void RadioButtons::OnApplyTemplate()
 {
-    winrt::IControlProtected controlProtected{ *this };
-    
-    m_listView.set(GetTemplateChildT<winrt::ListView>(L"InnerListView", controlProtected));
-    if (auto listView = m_listView.get())
+    const winrt::IControlProtected controlProtected{ *this };
+
+    m_repeater.set([this, controlProtected]() {
+        if(auto const repeater = GetTemplateChildT<winrt::ItemsRepeater>(s_repeaterName, controlProtected))
+        {
+            m_repeaterElementPreparedRevoker = repeater.ElementPrepared(winrt::auto_revoke, { this, &RadioButtons::OnRepeaterElementPrepared });
+            m_repeaterElementClearingRevoker = repeater.ElementClearing(winrt::auto_revoke, { this, &RadioButtons::OnRepeaterElementClearing });
+            m_repeaterElementIndexChangedRevoker = repeater.ElementIndexChanged(winrt::auto_revoke, { this, &RadioButtons::OnRepeaterElementIndexChanged });
+            m_repeaterLoadedRevoker = repeater.Loaded(winrt::auto_revoke, { this, &RadioButtons::OnRepeaterLoaded });
+            return repeater;
+        }
+        return static_cast<winrt::ItemsRepeater>(nullptr);
+    }());
+
+    UpdateItemsSource();
+}
+
+// When focus comes from outside the RadioButtons control we will put focus on the selected radio button.
+void RadioButtons::OnGettingFocus(const winrt::IInspectable&, const winrt::GettingFocusEventArgs& args)
+{
+    auto const inputDevice = args.InputDevice();
+    if (inputDevice == winrt::FocusInputDeviceKind::Keyboard || inputDevice == winrt::FocusInputDeviceKind::GameController)
     {
-        m_listViewLoadedRevoker = listView.Loaded(winrt::auto_revoke, { this, &RadioButtons::OnListViewLoaded });
-        m_listViewSelectionChangedRevoker = listView.SelectionChanged(winrt::auto_revoke, { this, &RadioButtons::OnListViewSelectionChanged });
+        if (auto const selectedIndex = m_selectionModel.SelectedIndex())
+        {
+            if (selectedIndex.GetSize() == 1)
```

Comment:
Is this check for size needed ?

Reply:
It could probably be assumed... I've removed selection model though.

---
In reply to: [347498055](https://github.com/microsoft/microsoft-ui-xaml/pull/1623#discussion_r347498055) [](ancestors = 347498055)